### ISSUE TITLE
Ocean Token reconciliation for Tethys v1.0

### ISSUE THREAD

**Issue by aaitor** [oceanprotocol/keeper-contracts#303]
- Ocean Token must live in `Pacific` and will be bridged to Ethereum mainnet. It means should be deployed independently of other keeper contracts. So we have 2 options (ask Sebastian):
  - Separate the Ocean Token in a different repository
  - Make the deployment script flexible enough to deploy different components like `OceanToken` or `MultiSigWallet` independently.

- During the initial (genesis) phase: 
  - All the balance are able to be used by the users once they received the token
  - The balance that is not able to be used because of a vesting period, is migrated by OPF in a further distribution. We are not having vesting business logic on-chain. The balance available in the `OceanToken` is ready to be used/transferred by the users.
  - This is going to be controlled by an external software releasing the tokens when the vesting period is over (out of keeper contracts) (Chalid)

- Different responsibilities will be assigned to different roles in the system. Each role will have the capability to interact with keeper contracts:
  - Contract Deployer: Can deploy contracts. In the keeper-contracts is managed using ZeppelinOS with MultiSig
  - Contract Upgrader (admin in `zos` terminology): Can upgrade contracts. In the keeper-contracts is managed using ZeppelinOS with MultiSig
  - Contract Administrator: Can modify the values of some attributes of the Smart Contracts, kill and stop them.
  - Minter Role: Have control over the minting operation to mint the Genesis supply. Will have the ability to mint all the 535,800,000 tokens from the initial allocation. The distribution is out of keeper contracts (Chalid).
  - Rewards Role: Have the capabilities of distributing the rewards (874,200,000) between different accounts of the system. Initially is not going to have any business logic.
  - Bridge Operator: In charge of mirroring or migrating the Ocean Tokens between Ethereum Mainnet and `Pacific`
  - Node Operator: Have the capability of managing the physical infrastructure. It means adding or removing nodes to the POA network, etc..
 

### Architecture

Architecture definition can be found here:
https://docs.google.com/presentation/d/14FmmLn4mCMptxAarbl04WmqRi6OuSB4Bv72Dd8wlp1Y/edit?usp=sharing
https://docs.google.com/presentation/d/1v8j-SwQM_APfXZfC1OotwxiXOAnp3j9482_iVXDcIO8/edit#slide=id.g3a22fb4e86_0_94

### Tasks

Define the tasks using as input the requirements & architecture

- [x] Formalize the specs/architecture of the Ocean Token 
- [x] Validate if the current interface of the Token is aligned
- [x] Setup the workflow. Converting the specs to tests in plain English
- [x] Formalize the tests in the keeper contracts project using the current testing framework
- [x] Implementation of the pending OceanToken.sol business logic (#304)
- [x] Implementation of the pending Dispenser.sol business logic (#305)
- [x] Implementation of the pending GenesisDispenser.sol business logic (#306)
- [x] Review of MultiSigWallet.sol business logic (#307)
- [x] Run the Unit & End to End tests
- [x] Separate Ocean Token from the existing contracts deployment (using an independent migration script)
- [x] Validate everything works okay with ZeppelinOS (understanding how to validate this from outside the UT)
- [x] Deploy the Ocean Token into Nile and run Acceptance Tests
- [x] Validate in `Nile` that after deploy the smart contracts are handover to the MultiSig
- [x] Test the process of really upgrading a contract in `Nile` at least once

**Comment by H34D** [comment]
> * Ocean Token must live in Ethereum mainnet and Pacific POA. It means should be deployed independently of other keeper contracts. So we have 2 options (ask Sebastian):

from the current planing this is not 100% precise, `OceanToken` will live in `pacific` and will be bridged to Ethereum mainnet using the token bridge. The bridge itself will deploy a contract in the ERC677 format in mainnet. https://github.com/poanetwork/poa-bridge-contracts/blob/master/contracts/ERC677.sol

**Comment by H34D** [comment]
> Ocean Token must live in Ethereum mainnet and Pacific POA. It means should be deployed independently of other keeper contracts. So we have 2 options (ask Sebastian):
> 
> * Separate the Ocean Token in a different repository
> * Implement an additional migration script deploying only the Ocean Token

we should make the deploy script so flexible to be able to only deploy `OceanToken` and `MultiSigWallet`

**Comment by H34D** [comment]
> * The balance not able to be used because a vesting period, is migrated by OPF in a further distribution. We are not having vesting business logic on-chain. The balance available in the OceanToken is ready to be used/transfered by the users.

we could `just not mint them` and leave them as mintable potential in the token rather than transferring them somewhere.

**Comment by H34D** [comment]
> * Genesis Role: Have control over the Genesis Dispenser. Needs to be registered on-chain (TODO). Will have the ability to mint all the 535,800,000 tokens from the initial allocation. This is out of keeper contracts (Chalid). The Genesis Dispenser lives on-chain.

we should do this a complete manual process because it is not reoccured. I described this in the corresponding user story.

**Comment by H34D** [comment]
Tested the governance stuff in `kovan` instead of `nile`.
